[Clang][Driver] Add an option to control loop-interchange

sjoerdmeijer opened the pull request:
This introduces options `-floop-interchange` and `-fno-loop-interchange` to enable/disable the loop-interchange pass. This is part of the work that tries to get that pass enabled by default (#124911), where it was remarked that a user facing option to control this would be convenient to have. The option name is the same as GCC's.

Meinersbur commented:
The other optimzation pass options (unrolll, vectorize, ...) are implemented in `PipelineTuningOptions` and `CodeGenOptions.def`. Do it the same way?

sjoerdmeijer committed:
[Clang][Driver] Add an option to control loop-interchange

This introduces options -floop-interchange and -fno-loop-interchange to
enable/disable the loop-interchange pass. This is part of the work that
tries to get that pass enabled by default (#124911), where it was
remarked that a user facing option to control this would be convenient
to have. The option (name) is the same as GCC's.

sjoerdmeijer commented:
> The other optimzation pass options (unrolll, vectorize, ...) are implemented in `PipelineTuningOptions` and `CodeGenOptions.def`. Do it the same way?

Thanks for the review and suggestion @Meinersbur, that's now implemented in the latest revision.

Meinersbur commented on llvm/lib/Passes/PassBuilderPipelines.cpp:
```diff
@@ -316,6 +312,7 @@ PipelineTuningOptions::PipelineTuningOptions() {
   LoopVectorization = true;
   SLPVectorization = false;
   LoopUnrolling = true;
+  LoopInterchange = false;
```
The default of `LoopInterchange` could be initialized with `EnableLoopInterchange`, as done by `MergeFunctions = EnableMergeFunction`.

Not important for me, but someone might want to keep `-mllvm -enable-loopinterchange` working, e.g. for non-Clang drivers.

sjoerdmeijer replied:
Good idea, that's now fixed.

madhur13490 reviewed (approved):
LGTM. Considering code-formatting issues will be fixed before committing.

Meinersbur reviewed (approved):
LGTM

Dont forget clang-format (https://github.com/llvm/llvm-project/pull/125830#issuecomment-2639658025)

sjoerdmeijer commented:
> This is a bit off topic, but do you have any opinion on adding a pragma for interchange like other loop optimizations do? I think it can sometimes be useful if we can enable/disable the interchange for each loop, but I think there are a few things to consider if we introduce it, such as the behavior when the pragma is specified on a non-outermost loop. What do you think?

Yeah, I see the use case for that, I think that makes sense. And OpenMP6 for example, introduces a directive for interchange. Please propose a patch if you're interested in pursuing this.

Meinersbur commented:
The OpenMP pragma conceptually applies interchange in the frontend, with syntactical requirements for the loops it applies to.

kasuga-fj commented:
Thanks for the comments. I'll try to work on it in the near future.
